### ISSUE TITLE
Strange key fetching behaviour

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#12293]
I have a test case demonstrating some unexpected behaviour from `EVP_KEYMGMT_fetch`. The test loads two providers: legacy and the tls-provider (from test/tls-provider.c). It then attempts the fetch the "XOR" keymgmt which is defined in the tls-provider (legacy has no keymgmt functions).

Dependent on the order that I load the providers in the fetch succeeds or fails. I would expect it to succeed regardless of the order.

````
$ ./tmptest
$ ./tmptest -swap
tmptest: test/tmptest.c:35: main: Assertion `xorkeymgmt != NULL' failed.
Aborted (core dumped)
````

Full testcase below

```` diff
From 6277a3899b6e895430bc3217a205dd1dc867259d Mon Sep 17 00:00:00 2001
From: Matt Caswell <matt@openssl.org>
Date: Sat, 27 Jun 2020 00:20:40 +0100
Subject: [PATCH] Demonstrate strange behaviour fetching a key

---
 test/build.info |  6 +++++-
 test/tmptest.c  | 38 ++++++++++++++++++++++++++++++++++++++
 2 files changed, 43 insertions(+), 1 deletion(-)
 create mode 100644 test/tmptest.c

diff --git a/test/build.info b/test/build.info
index 88b35d4d3c..cc1f39afa8 100644
--- a/test/build.info
+++ b/test/build.info
@@ -57,7 +57,7 @@ IF[{- !$disabled{tests} -}]
           http_test servername_test ocspapitest fatalerrtest tls13ccstest \
           sysdefaulttest errtest ssl_ctx_test gosttest \
           context_internal_test aesgcmtest params_test evp_pkey_dparams_test \
-          keymgmt_internal_test hexstr_test
+          keymgmt_internal_test hexstr_test tmptest
 
   SOURCE[confdump]=confdump.c
   INCLUDE[confdump]=../include ../apps/include
@@ -307,6 +307,10 @@ IF[{- !$disabled{tests} -}]
   INCLUDE[sslapitest]=../include ../apps/include ..
   DEPEND[sslapitest]=../libcrypto ../libssl libtestutil.a
 
+  SOURCE[tmptest]=tmptest.c filterprov.c tls-provider.c
+  INCLUDE[tmptest]=../include ../apps/include ..
+  DEPEND[tmptest]=../libcrypto libtestutil.a
+
   SOURCE[ocspapitest]=ocspapitest.c
   INCLUDE[ocspapitest]=../include ../apps/include
   DEPEND[ocspapitest]=../libcrypto libtestutil.a
diff --git a/test/tmptest.c b/test/tmptest.c
new file mode 100644
index 0000000000..c0563e86fd
--- /dev/null
+++ b/test/tmptest.c
@@ -0,0 +1,38 @@
+#include <assert.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/provider.h>
+
+
+/* Defined in tls-provider.c */
+int tls_provider_init(const OSSL_CORE_HANDLE *handle,
+                      const OSSL_DISPATCH *in,
+                      const OSSL_DISPATCH **out,
+                      void **provctx);
+
+int main(int argc, char *argv[])
+{
+    OPENSSL_CTX *libctx = NULL;
+    EVP_KEYMGMT *xorkeymgmt = NULL;
+
+    libctx = OPENSSL_CTX_new();
+    assert(libctx != NULL);
+
+    OSSL_PROVIDER_load(NULL, "null");
+
+    assert(OSSL_PROVIDER_add_builtin(libctx, "tls-provider", tls_provider_init));
+
+    if (argc == 2 && strcmp(argv[1], "-swap") == 0) {
+        printf("Swapping");
+        OSSL_PROVIDER_load(libctx, "legacy");
+        OSSL_PROVIDER_load(libctx, "tls-provider");
+    } else {
+        OSSL_PROVIDER_load(libctx, "tls-provider");
+        OSSL_PROVIDER_load(libctx, "legacy");
+    }
+
+    xorkeymgmt = EVP_KEYMGMT_fetch(libctx, "XOR", NULL);
+    assert(xorkeymgmt != NULL);
+
+    return 0;
+}
-- 
2.25.1
````

**Comment by levitte** [comment]
That is indeed strange... Will look.

**Comment by levitte** [comment]
Is this solved by #12292 by any chance?

**Comment by mattcaswell** [comment]
> Is this solved by #12292 by any chance?

No. It was while writing a test for #12292 that I ran into this. I had to adjust my test to avoid it.

**Comment by mattcaswell** [comment]
Any thoughts on this @levitte? Looks like #12292 is now blocked on this

**Comment by levitte** [comment]
It turns out that the core issue is here:

https://github.com/openssl/openssl/blob/64bb6276d17cc78c15e0bbea2cce899ba9b6778d/crypto/core_algorithm.c#L61-L62

If the operation query returned NULL and this was a single operation, this effectively returns from the function with 0 (because `ok = 0`), which stops any further attempt to query the operation in any following providers.  Since the "tls-provider" came after "legacy" in the provider store in the swapped case, and "legacy" doesn't have any KEYMGMT map, the search for "XOR" stops short.

**Comment by levitte** [comment]
Fix in #12365, please try it out.
